bridge-history-api/internal/logic/history_logic.go

```diff
@@ -278,36 +278,41 @@ func getTxHistoryInfo(message *orm.CrossMessage) *types.TxHistoryInfo {
 				Message:    message.MessageData,
 				Proof:      common.Bytes2Hex(message.MerkleProof),
 				BatchIndex: strconv.FormatUint(message.BatchIndex, 10),
+				Claimable:  true,
 			}
 		}
 	}
 	return txHistory
 }
 
 func (h *HistoryLogic) getCachedTxsInfo(ctx context.Context, cacheKey string, pageNum, pageSize uint64) ([]*types.TxHistoryInfo, uint64, bool, error) {
-	start := int64(pageNum * pageSize)
-	end := int64((pageNum+1)*pageSize - 1)
+	start := int64((pageNum - 1) * pageSize)
+	end := int64(pageNum*pageSize - 1)
 
 	total, err := h.redis.ZCard(ctx, cacheKey).Result()
 	if err != nil {
-		if err == redis.Nil {
-			// Key does not exist, cache miss.
-			return nil, 0, false, nil
-		}
 		log.Error("failed to get zcard result", "error", err)
 		return nil, 0, false, err
 	}
 
 	values, err := h.redis.ZRange(ctx, cacheKey, start, end).Result()
 	if err != nil {
-		if err == redis.Nil {
-			// Key does not exist, cache miss.
-			return nil, 0, false, nil
-		}
 		log.Error("failed to get zrange result", "error", err)
 		return nil, 0, false, err
 	}
 
+	// Perform a final check to confirm the existence of the key to ensure consistency between ZCard and ZRange data reads.
+	exists, err := h.redis.Exists(ctx, cacheKey).Result()
+	if err != nil {
+		log.Error("failed to check if key exists", "error", err)
+		return nil, 0, false, err
+	}
+
+	// If the key does not exist, we consider it a cache miss and return accordingly.
+	if exists == 0 {
+		return nil, 0, false, nil
+	}
```

Comment:
it's necessary because if the key does not exist, it will return nil value, and cannot differentiate:
1. cache key exists but the value is nil.
2. cache key does not exist.

Reply:
i think just check total or value whether empty

Reply:
Yeah. but if the txs under the address are empty, it would be treated as "key not found", too.

Reply:
changed in [1d325cb](https://github.com/scroll-tech/scroll/pull/1029/commits/1d325cb54bf65b09abda141a58b1d1fe9dbd43f3)